<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="metamoRph">
<title>So many ways to screw up PCA projection • metamoRph</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="So many ways to screw up PCA projection">
<meta property="og:description" content="metamoRph">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">metamoRph</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.2.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/So_many_ways_to_screw_up_PCA_projection.html">So many ways to screw up PCA projection</a>
    <a class="dropdown-item" href="../articles/real_data_testing.html">Metadata Label Transfer on Real Data</a>
    <a class="dropdown-item" href="../articles/build_a_brain_region_predictor.html">Build and Apply a Human Brain Region Predictor in 60 seconds or less</a>
    <a class="dropdown-item" href="../articles/morph_scRNA_query_onto_reference.html">Transfer Labels from a Retina scRNA Reference onto a Query Dataset</a>
    <a class="dropdown-item" href="../articles/dataset_integration_and_projection.html">Batch Corrected metamoRph Reference</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>So many ways to screw up PCA projection</h1>
            
            <h4 data-toc-skip class="date">2023-07-25</h4>
      
      
      <div class="d-none name"><code>So_many_ways_to_screw_up_PCA_projection.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>When you run <code>prcomp</code> you get two matrices back:
<code>$x</code> and <code>$rotation</code>. The former is the “pattern
matrix” which are the factors (principal components) by samples. The
latter is the amplitude matrix, which are the loadings of the features
(genes) with the factors (principal compoent); row by column,
respectively.</p>
<p>In <em>theory</em> you can use a rotation (loadings) matrix from
prcomp (the <code>$rotation</code> slot) to multiply against a new
matrix to get the new matrix in the same eigenvector space as the
original matrix.</p>
<p>This is handy! This means you can use an existing PCA you understand
and have spent time interpreting and transfer that knowledge onto new
data.</p>
<p>In <em>practice</em>, though, there a wide variety of reasons this
can fail. Let us explore why.</p>
</div>
<div class="section level2">
<h2 id="toy-data">Toy Data<a class="anchor" aria-label="anchor" href="#toy-data"></a>
</h2>
<p>Let’s make a tiny little faux dataset. Threetissue types with five
genes. We <code>log1p</code> scale the data as raw gene counts are
skewed, run <code>prcomp</code>, then plot the first two principal
components (PC). They do a nice job turning three major axes of
variation into two informative eigenvalues/PCs.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'RHO'</span>,<span class="st">'RPE65'</span>,<span class="st">'ABCA4'</span>,<span class="st">'PMEL'</span>, <span class="st">'KRT5'</span><span class="op">)</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Retina1'</span>,<span class="st">'Retina2'</span>,<span class="st">'RPE1'</span>,<span class="st">'RPE2'</span>,<span class="st">'Cornea1'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">faux_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">560</span>,<span class="fl">650</span>,<span class="fl">5</span>,<span class="fl">6</span>,<span class="fl">9</span><span class="op">)</span>, <span class="co"># rho</span></span>
<span>                  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">42</span>,<span class="fl">32</span>, <span class="fl">1103</span>,<span class="fl">1201</span>,<span class="fl">2</span><span class="op">)</span>, <span class="co">#rpe65</span></span>
<span>                  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">810</span>,<span class="fl">903</span>,<span class="fl">202</span>,<span class="fl">205</span>,<span class="fl">45</span><span class="op">)</span>,  <span class="co">#abca4</span></span>
<span>                  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">100</span>,<span class="fl">105</span>,<span class="fl">2004</span>,<span class="fl">1980</span>,<span class="fl">101</span><span class="op">)</span>,<span class="co"># pmel</span></span>
<span>                  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">32</span>,<span class="fl">101</span>,<span class="fl">202</span>,<span class="fl">1867</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="op">)</span><span class="co"># krt5</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">faux_mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">genes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">faux_mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">samples</span></span>
<span></span>
<span><span class="va">faux_pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p</a></span><span class="op">(</span><span class="va">faux_mat</span><span class="op">)</span>, scale <span class="op">=</span> <span class="cn">TRUE</span>, center <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">faux_pca</span><span class="op">$</span><span class="va">x</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">'samples'</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">'\\d'</span>,<span class="st">''</span>,<span class="va">samples</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">PC1</span>,y<span class="op">=</span><span class="va">PC2</span>,color<span class="op">=</span><span class="va">samples</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/theme_cowplot.html" class="external-link">theme_cowplot</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="So_many_ways_to_screw_up_PCA_projection_files/figure-html/unnamed-chunk-2-1.png" width="384"></p>
<div class="section level3">
<h3 id="matrix-multiplication">Matrix Multiplication<a class="anchor" aria-label="anchor" href="#matrix-multiplication"></a>
</h3>
<p>The pattern matrix (plotted above) can be recreated by multiplying
the input_matrix by the rotation matrix (the <code>$rotation</code>
slot).</p>
<p>Wait, that’s wrong….the numbers are very different. PC1 should range
from about -2 to 2, but now runs from -5 to 3.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p</a></span><span class="op">(</span><span class="va">faux_mat</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="va">faux_pca</span><span class="op">$</span><span class="va">rotation</span></span>
<span><span class="co">#&gt;               PC1      PC2       PC3         PC4      PC5</span></span>
<span><span class="co">#&gt; Retina1 -3.665190 6.706788 -7.344774 -0.19334184 3.146782</span></span>
<span><span class="co">#&gt; Retina2 -2.829324 5.805728 -9.090433 -0.17940189 3.181682</span></span>
<span><span class="co">#&gt; RPE1     3.019202 8.865143 -8.318996 -0.03906449 1.680128</span></span>
<span><span class="co">#&gt; RPE2     3.263775 8.661185 -8.930563 -0.21079129 1.690195</span></span>
<span><span class="co">#&gt; Cornea1  2.472395 1.693552 -9.136326  1.37485738 2.225891</span></span></code></pre></div>
<p>What happened? If we investigate the <code>prcomp</code>
documentation we see that the input data, by default, is scaled across
the features. If we apply the <code>scale</code> function to our log1p
scaled <code>faux_mat</code> we get a matrix which exactly matches the
prcomp <code>$x</code> slot.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p</a></span><span class="op">(</span><span class="va">faux_mat</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="va">faux_pca</span><span class="op">$</span><span class="va">rotation</span></span>
<span><span class="co">#&gt;               PC1        PC2         PC3           PC4           PC5</span></span>
<span><span class="co">#&gt; Retina1 -2.057963  0.2320273  0.36944476 -0.0116954785  8.204038e-16</span></span>
<span><span class="co">#&gt; Retina2 -1.701178 -0.1366033 -0.42140684  0.0118873413  4.905776e-16</span></span>
<span><span class="co">#&gt; RPE1     1.254511  1.1468746  0.12984655  0.0347133588 -6.470796e-17</span></span>
<span><span class="co">#&gt; RPE2     1.360317  1.0552042 -0.14052654 -0.0345002389 -4.941149e-17</span></span>
<span><span class="co">#&gt; Cornea1  1.144313 -2.2975028  0.06264207 -0.0004049826 -4.716018e-17</span></span>
<span></span>
<span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p</a></span><span class="op">(</span><span class="va">faux_mat</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="va">faux_pca</span><span class="op">$</span><span class="va">rotation</span><span class="op">)</span> <span class="op">==</span> <span class="va">faux_pca</span><span class="op">$</span><span class="va">x</span></span>
<span><span class="co">#&gt;          PC1  PC2  PC3  PC4  PC5</span></span>
<span><span class="co">#&gt; Retina1 TRUE TRUE TRUE TRUE TRUE</span></span>
<span><span class="co">#&gt; Retina2 TRUE TRUE TRUE TRUE TRUE</span></span>
<span><span class="co">#&gt; RPE1    TRUE TRUE TRUE TRUE TRUE</span></span>
<span><span class="co">#&gt; RPE2    TRUE TRUE TRUE TRUE TRUE</span></span>
<span><span class="co">#&gt; Cornea1 TRUE TRUE TRUE TRUE TRUE</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="three-fake-cornea-tissues">Three fake cornea tissues<a class="anchor" aria-label="anchor" href="#three-fake-cornea-tissues"></a>
</h3>
<p>Now we have new data. Three cornea samples. We want to compare these
new cornea samples to our existing PCA to see how they compare. As
discussed above, we <em>should</em> be able to just matrix multiply this
new matrix against the original pca <code>$loading</code> slot to put
the new data in the same PC space.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">new_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>,<span class="fl">6</span>,<span class="fl">3</span><span class="op">)</span>, <span class="co"># rho</span></span>
<span>                  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">7</span>,<span class="fl">2</span>,<span class="fl">6</span><span class="op">)</span>, <span class="co">#rpe65</span></span>
<span>                  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">65</span>,<span class="fl">22</span>,<span class="fl">10</span><span class="op">)</span>,  <span class="co">#abca4</span></span>
<span>                  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">122</span>,<span class="fl">101</span>,<span class="fl">57</span><span class="op">)</span>,<span class="co"># pmel</span></span>
<span>                  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2567</span>,<span class="fl">1755</span>,<span class="fl">2218</span><span class="op">)</span><span class="op">)</span> <span class="co"># krt5</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">new_data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">genes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">new_data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cornea2"</span>,<span class="st">"Cornea3"</span>,<span class="st">"Cornea4"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="ok-lets-try-projecting-some-new-data">OK, let’s try projecting some new data!<a class="anchor" aria-label="anchor" href="#ok-lets-try-projecting-some-new-data"></a>
</h2>
<p>We apply all of knowledge and scale the new data before we matrix
multiply it against the rotation matrix.</p>
<p>What the? The three new samples are … in the middle??? Why??????</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">new_project</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p</a></span><span class="op">(</span><span class="va">new_data</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="va">faux_pca</span><span class="op">$</span><span class="va">rotation</span></span>
<span></span>
<span><span class="co"># we bind rows to glue together the original PCA (faux_pca$x)</span></span>
<span><span class="co"># with the new data's PC space (new_project)</span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="va">faux_pca</span><span class="op">$</span><span class="va">x</span>, rownames <span class="op">=</span> <span class="st">'samples'</span><span class="op">)</span>,</span>
<span>      <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="va">new_project</span>, rownames <span class="op">=</span> <span class="st">'samples'</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">'\\d'</span>,<span class="st">''</span>,<span class="va">samples</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">PC1</span>,y<span class="op">=</span><span class="va">PC2</span>,color<span class="op">=</span><span class="va">samples</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/theme_cowplot.html" class="external-link">theme_cowplot</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="So_many_ways_to_screw_up_PCA_projection_files/figure-html/unnamed-chunk-6-1.png" width="384"></p>
</div>
<div class="section level2">
<h2 id="what-scale-does">What scale does<a class="anchor" aria-label="anchor" href="#what-scale-does"></a>
</h2>
<p><code>scale</code> takes the columns and, by default, will center
(mean 0) and scale (divide by the standard deviation).</p>
<p>The issue is that because you only have <em>one</em> sample type with
fairly <em>similar</em> expression patterns across all the genes, the
scaling turns <em>everything</em> into similar values. Which centers
them around 0, in effect when multiplying by the <code>prcomp</code>
rotation matrix.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="va">new_data</span><span class="op">)</span></span>
<span><span class="co">#&gt;                RHO      RPE65      ABCA4       PMEL       KRT5</span></span>
<span><span class="co">#&gt; Cornea2  0.2182179  0.7559289  1.1295751  0.8642016  0.9500860</span></span>
<span><span class="co">#&gt; Cornea3  0.8728716 -1.1338934 -0.3573146  0.2311237 -1.0433761</span></span>
<span><span class="co">#&gt; Cornea4 -1.0910895  0.3779645 -0.7722605 -1.0953253  0.0932901</span></span>
<span><span class="co">#&gt; attr(,"scaled:center")</span></span>
<span><span class="co">#&gt;         RHO       RPE65       ABCA4        PMEL        KRT5 </span></span>
<span><span class="co">#&gt;    4.666667    5.000000   32.333333   93.333333 2180.000000 </span></span>
<span><span class="co">#&gt; attr(,"scaled:scale")</span></span>
<span><span class="co">#&gt;        RHO      RPE65      ABCA4       PMEL       KRT5 </span></span>
<span><span class="co">#&gt;   1.527525   2.645751  28.919428  33.171273 407.331560</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="how-to-fix">How to fix?<a class="anchor" aria-label="anchor" href="#how-to-fix"></a>
</h2>
<div class="section level3">
<h3 id="with-statspredict">with stats::predict<a class="anchor" aria-label="anchor" href="#with-statspredict"></a>
</h3>
<p>If you use the <code><a href="https://rdrr.io/r/stats/predict.html" class="external-link">stats::predict</a></code> tool, then it will use the
scaling information in the <code>prcomp</code> object to scale your new
data with the same scale and center values used in in the
<code>prcomp</code> scale/center steps.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">with_predict</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">faux_pca</span>, <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p</a></span><span class="op">(</span><span class="va">new_data</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="va">faux_pca</span><span class="op">$</span><span class="va">x</span>, rownames <span class="op">=</span> <span class="st">'samples'</span><span class="op">)</span>,</span>
<span>      <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="va">with_predict</span>, rownames <span class="op">=</span> <span class="st">'samples'</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">'\\d'</span>,<span class="st">''</span>,<span class="va">samples</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">PC1</span>,y<span class="op">=</span><span class="va">PC2</span>,color<span class="op">=</span><span class="va">samples</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/theme_cowplot.html" class="external-link">theme_cowplot</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="So_many_ways_to_screw_up_PCA_projection_files/figure-html/unnamed-chunk-8-1.png" width="384"></p>
</div>
</div>
<div class="section level2">
<h2 id="yay-all-is-well-in-the-world">Yay all is well in the world!<a class="anchor" aria-label="anchor" href="#yay-all-is-well-in-the-world"></a>
</h2>
<div class="section level3">
<h3 id="plot-twist">plot twist<a class="anchor" aria-label="anchor" href="#plot-twist"></a>
</h3>
<p>In RNA-seq data it is common to have wildly different sum counts.
That is because a sequencing machine can create 10,000,000 reads for one
sample or 100,000,000 reads. Or even more!</p>
<p>Let’s try <code><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">stats::prcomp</a></code> again with the same new_data,
just scaled up 20x.</p>
<p>Ugh, now the new cornea samples look more like the RPE.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">with_predict</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">faux_pca</span>, <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p</a></span><span class="op">(</span><span class="va">new_data</span><span class="op">*</span><span class="fl">20</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="va">faux_pca</span><span class="op">$</span><span class="va">x</span>, rownames <span class="op">=</span> <span class="st">'samples'</span><span class="op">)</span>,</span>
<span>      <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="va">with_predict</span>, rownames <span class="op">=</span> <span class="st">'samples'</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">'\\d'</span>,<span class="st">''</span>,<span class="va">samples</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">PC1</span>,y<span class="op">=</span><span class="va">PC2</span>,color<span class="op">=</span><span class="va">samples</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/theme_cowplot.html" class="external-link">theme_cowplot</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="So_many_ways_to_screw_up_PCA_projection_files/figure-html/unnamed-chunk-9-1.png" width="384"></p>
</div>
<div class="section level3">
<h3 id="ok-so-this-is-fixable-">OK, so this is fixable….<a class="anchor" aria-label="anchor" href="#ok-so-this-is-fixable-"></a>
</h3>
<p>You can either re-do the original PCA with a fixed sample scale (like
cpm) and/or alter the new data to have a similar scale to the input data
PCA.</p>
<p>But to recap, at this point we have discussed several issues that can
make data projection challenging:</p>
<ol style="list-style-type: decimal">
<li>log scaling of input data</li>
<li>log scaling of new data</li>
<li>row scaling (feature/genes) within prcomp needs to be applied in the
same way to the new data</li>
<li>unequal sum counts of the samples need to be normalized either at
the first <code>prcomp</code> step or by adjusting the new data</li>
</ol>
<p>There are even <em>more</em> little annoyances that I have not
discussed:</p>
<ol start="5" style="list-style-type: decimal">
<li>highly variable gene selection is usually done so you don’t do PCA
on 20,000+ genes.</li>
</ol>
<ul>
<li>this means that the new data needs to be exactly matched at the
feature/gene level to whatever the highly variable genes used in the
prcomp step</li>
</ul>
<ol start="6" style="list-style-type: decimal">
<li>if you are using a new a RNA seq dataset from a different
quantification setup than the original data the gene names can be
slightly different, which means you may have a handful of missing
features/genes that have to be addressed in some way otherwise the
matrix multiplication will fail…</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="wow-if-only-someone-wrote-a-r-package-that-could-handle-all-of-the-issues-so-i-can-get-on-with-my-analysis">wow, if only someone wrote a R package that could handle all of the
issues so I can get on with my analysis!<a class="anchor" aria-label="anchor" href="#wow-if-only-someone-wrote-a-r-package-that-could-handle-all-of-the-issues-so-i-can-get-on-with-my-analysis"></a>
</h2>
<div class="section level3">
<h3 id="good-news-i-did-metamorph">Good news, I did: <a href="www.github.com/davemcg/metamoRph">metamoRph</a><a class="anchor" aria-label="anchor" href="#good-news-i-did-metamorph"></a>
</h3>
<p>There are two steps:</p>
<ol style="list-style-type: decimal">
<li>A <code>run_pca</code> wrapper around <code>prcomp</code> which, by
default, will:</li>
</ol>
<ul>
<li>CPM scale the samples</li>
<li>log1p scale the data</li>
<li>remove zero count genes</li>
<li>row scale and center the genes</li>
<li>tries to remove mitochondrial and ribosomal genes which are often
not useful for scRNA analysis</li>
<li>offers two ways to select highly variable genes: “scran” which uses
a count scaled variance selection approach (higher expressed
genes/features will naturally have higher variance) or the “classic”
approach which just uses <code>matrixStats::rowVar</code>
</li>
<li>calculates the % standard deviation explained by each PC</li>
<li>returns a list object with the prcomp object, the parameters chosen,
the center/scale values, and the % standard deviation explained by each
PC</li>
</ul>
<ol start="2" style="list-style-type: decimal">
<li>
<code>metamoRph</code> function which takes two or three
inputs:</li>
</ol>
<ul>
<li>your new counts matrix</li>
<li>the <code>$rotation</code> matrix from <code>prcomp</code>
</li>
<li>optionally the center/scale values from <code>prcomp</code>
</li>
</ul>
<p>and then</p>
<ul>
<li>will align the feature/gene names from the input data to the
rotation matrix</li>
<li>scale the data in all directions discussed above</li>
<li>return the <code>$x</code> equivalent PC space</li>
</ul>
<p>Let’s see it in practice with the 10x scaled input data that tripped
up <code><a href="https://rdrr.io/r/stats/predict.html" class="external-link">stats::predict</a></code></p>
<p>There two only two steps:</p>
<ol style="list-style-type: decimal">
<li>
<code>run_pca</code> which runs the prcomp with the settings
discussed above</li>
<li>
<code>metamoRph</code> which takes the new data and projects it onto
the <code>mm_pca</code> space</li>
</ol>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://davemcg.github.io/metamoRph/">metamoRph</a></span><span class="op">)</span></span>
<span><span class="va">mm_pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pca.html">run_pca</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">faux_mat</span><span class="op">)</span>, meta <span class="op">=</span> <span class="va">samples</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="op">)</span>, sample_scale <span class="op">=</span> <span class="st">'seurat'</span><span class="op">)</span></span>
<span><span class="va">projected_pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/metamoRph.html">metamoRph</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">new_data</span><span class="op">*</span><span class="fl">10</span><span class="op">)</span>, </span>
<span>                               <span class="va">mm_pca</span><span class="op">$</span><span class="va">PCA</span><span class="op">$</span><span class="va">rotation</span>, </span>
<span>                               center_scale <span class="op">=</span> <span class="va">mm_pca</span><span class="op">$</span><span class="va">center_scale</span>, sample_scale <span class="op">=</span> <span class="st">'seurat'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="va">faux_pca</span><span class="op">$</span><span class="va">x</span>, rownames <span class="op">=</span> <span class="st">'samples'</span><span class="op">)</span>,</span>
<span>      <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="va">projected_pca</span>, rownames <span class="op">=</span> <span class="st">'samples'</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">'\\d'</span>,<span class="st">''</span>,<span class="va">samples</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">PC1</span>,y<span class="op">=</span><span class="va">PC2</span>,color<span class="op">=</span><span class="va">samples</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/theme_cowplot.html" class="external-link">theme_cowplot</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="So_many_ways_to_screw_up_PCA_projection_files/figure-html/unnamed-chunk-10-1.png" width="384"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">#&gt; Platform: aarch64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Ventura 13.4.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: America/New_York</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] metamoRph_0.2.0 ggplot2_3.4.2   dplyr_1.1.2    </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] tidyselect_1.2.0            farver_2.1.1               </span></span>
<span><span class="co">#&gt;  [3] bitops_1.0-7                fastmap_1.1.1              </span></span>
<span><span class="co">#&gt;  [5] SingleCellExperiment_1.22.0 RCurl_1.98-1.12            </span></span>
<span><span class="co">#&gt;  [7] bluster_1.9.1               digest_0.6.31              </span></span>
<span><span class="co">#&gt;  [9] rsvd_1.0.5                  lifecycle_1.0.3            </span></span>
<span><span class="co">#&gt; [11] cluster_2.1.4               statmod_1.5.0              </span></span>
<span><span class="co">#&gt; [13] magrittr_2.0.3              compiler_4.3.0             </span></span>
<span><span class="co">#&gt; [15] rlang_1.1.1                 sass_0.4.6                 </span></span>
<span><span class="co">#&gt; [17] tools_4.3.0                 igraph_1.4.3               </span></span>
<span><span class="co">#&gt; [19] utf8_1.2.3                  yaml_2.3.7                 </span></span>
<span><span class="co">#&gt; [21] knitr_1.43                  dqrng_0.3.0                </span></span>
<span><span class="co">#&gt; [23] S4Arrays_1.0.4              labeling_0.4.2             </span></span>
<span><span class="co">#&gt; [25] DelayedArray_0.26.6         BiocParallel_1.33.11       </span></span>
<span><span class="co">#&gt; [27] withr_2.5.0                 purrr_1.0.1                </span></span>
<span><span class="co">#&gt; [29] BiocGenerics_0.46.0         desc_1.4.2                 </span></span>
<span><span class="co">#&gt; [31] grid_4.3.0                  stats4_4.3.0               </span></span>
<span><span class="co">#&gt; [33] fansi_1.0.4                 beachmat_2.15.0            </span></span>
<span><span class="co">#&gt; [35] colorspace_2.1-0            edgeR_3.42.2               </span></span>
<span><span class="co">#&gt; [37] scales_1.2.1                SummarizedExperiment_1.30.1</span></span>
<span><span class="co">#&gt; [39] cli_3.6.1                   rmarkdown_2.21             </span></span>
<span><span class="co">#&gt; [41] crayon_1.5.2                ragg_1.2.5                 </span></span>
<span><span class="co">#&gt; [43] generics_0.1.3              metapod_1.7.0              </span></span>
<span><span class="co">#&gt; [45] rstudioapi_0.14             DelayedMatrixStats_1.21.0  </span></span>
<span><span class="co">#&gt; [47] scuttle_1.9.4               cachem_1.0.8               </span></span>
<span><span class="co">#&gt; [49] stringr_1.5.0               zlibbioc_1.46.0            </span></span>
<span><span class="co">#&gt; [51] parallel_4.3.0              XVector_0.40.0             </span></span>
<span><span class="co">#&gt; [53] matrixStats_0.63.0          vctrs_0.6.2                </span></span>
<span><span class="co">#&gt; [55] Matrix_1.5-4.1              jsonlite_1.8.4             </span></span>
<span><span class="co">#&gt; [57] BiocSingular_1.15.0         IRanges_2.34.0             </span></span>
<span><span class="co">#&gt; [59] S4Vectors_0.38.1            BiocNeighbors_1.17.1       </span></span>
<span><span class="co">#&gt; [61] irlba_2.3.5.1               systemfonts_1.0.4          </span></span>
<span><span class="co">#&gt; [63] locfit_1.5-9.7              limma_3.56.1               </span></span>
<span><span class="co">#&gt; [65] tidyr_1.3.0                 jquerylib_0.1.4            </span></span>
<span><span class="co">#&gt; [67] glue_1.6.2                  pkgdown_2.0.7              </span></span>
<span><span class="co">#&gt; [69] codetools_0.2-19            cowplot_1.1.1              </span></span>
<span><span class="co">#&gt; [71] stringi_1.7.12              gtable_0.3.3               </span></span>
<span><span class="co">#&gt; [73] GenomeInfoDb_1.36.0         GenomicRanges_1.52.0       </span></span>
<span><span class="co">#&gt; [75] ScaledMatrix_1.8.1          munsell_0.5.0              </span></span>
<span><span class="co">#&gt; [77] tibble_3.2.1                pillar_1.9.0               </span></span>
<span><span class="co">#&gt; [79] htmltools_0.5.5             GenomeInfoDbData_1.2.10    </span></span>
<span><span class="co">#&gt; [81] R6_2.5.1                    textshaping_0.3.6          </span></span>
<span><span class="co">#&gt; [83] sparseMatrixStats_1.11.1    rprojroot_2.0.3            </span></span>
<span><span class="co">#&gt; [85] evaluate_0.21               lattice_0.21-8             </span></span>
<span><span class="co">#&gt; [87] Biobase_2.60.0              highr_0.10                 </span></span>
<span><span class="co">#&gt; [89] memoise_2.0.1               scran_1.27.1               </span></span>
<span><span class="co">#&gt; [91] bslib_0.5.0                 Rcpp_1.0.10                </span></span>
<span><span class="co">#&gt; [93] xfun_0.39                   fs_1.6.2                   </span></span>
<span><span class="co">#&gt; [95] MatrixGenerics_1.12.0       pkgconfig_2.0.3</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by David McGaughey.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
